system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+ * Name: get_setting
+ *
+ * Description:
+ *    Gets a setting for a string value
+ *
+ * Input Parameters:
+ *    key        - key of the required setting
+ *    setting    - pointer to pointer for the setting
+ *
+ * Returned Value:
+ *   The value of the setting for the given key
+ *
+ ****************************************************************************/
+
+static int get_setting(FAR char *key, FAR setting_t **setting)
+{
+  DEBUGASSERT(*setting == NULL);
+
+  int i;
+  int ret = OK;
+
+  for (i = 0; i < CONFIG_SYSTEM_SETTINGS_MAP_SIZE; i++)
+    {
+      if (map[i].type == SETTING_EMPTY)
+        {
+          ret = -ENOENT;
+          goto exit;
+        }
+
+      if (strcmp(map[i].key, key) == 0)
+        {
+          *setting = &map[i];
+          goto exit;
+        }
+    }
+
+  ret = -ENOENT;
+
+exit:
+  return ret;
+}
+
+/****************************************************************************
+ * Name: get_string
+ *
+ * Description:
+ *    Gets a setting for a string value
+ *
+ * Input Parameters:
+ *    setting        - pointer to the setting type
+ *    str            - pointer to return the string value
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static size_t get_string(FAR setting_t *setting, FAR char *buffer,
+                         size_t size)
+{
+  DEBUGASSERT(setting);
+  DEBUGASSERT((setting->type == SETTING_STRING) ||
+              (setting->type == SETTING_IP_ADDR));
+
+  if (setting->type == SETTING_STRING)
+    {
+      FAR const char *s = setting->val.s;
+      size_t len = strlen(s);
+
+      DEBUGASSERT(len < size);
+      if (len >= size)
+        {
+          return 0;
+        }
+
+      strncpy(buffer, s, size);
+      buffer[size - 1] = '\0';
+
+      return len;
+    }
+  else if (setting->type == SETTING_IP_ADDR)
+    {
+      DEBUGASSERT(INET_ADDRSTRLEN < size);
+
+      inet_ntop(AF_INET, &setting->val.ip, buffer, size);
+      buffer[size - 1] = '\0';
+
+      return strlen(buffer);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: set_string
+ *
+ * Description:
+ *    Creates a setting for a string valuesetting->type != set_
+ *
+ * Input Parameters:
+ *    setting        - pointer to the setting type
+ *    str            - the string value
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int set_string(FAR setting_t *setting, FAR char *str)
+{
+  DEBUGASSERT(setting);
+
+  if ((setting->type != SETTING_STRING) &&
+              (setting->type != SETTING_IP_ADDR))
+    {
+      return -EACCES;
+    }
+
+  ASSERT(strlen(str) < CONFIG_SYSTEM_SETTINGS_VALUE_SIZE);
+  if (strlen(str) >= CONFIG_SYSTEM_SETTINGS_VALUE_SIZE)
+    {
+      return -EINVAL;
+    }
+
+  if (strlen(str) && (sanity_check(str) < 0))
```

Comment:
This check fails when `strlen(str) == 0`.

Reply:
yep, (il)logical error!!

Reply:
Oops, you are right! The code allows for empty strings, and this check only guards the sanity check.